src/controller/InvokeInteraction.h

```diff
@@ -30,29 +30,31 @@ namespace Controller {
  * the provided success callback or calls the provided failure callback.
  *
  * The RequestObjectT is generally expected to be a ClusterName::Commands::CommandName::Type struct, but any object
- * that can be encoded using the DataModel::Encode machinery and exposes the GetClusterId() and GetCommandId() functions
+ * that can be encoded using the DataModel::Encode machinery and exposes the
+ * GetClusterId() and GetCommandId() functions and a ResponseType type
  * is expected to work.
  *
- * The ResponseObjectT is expected to be one of two things:
+ * The ResponseType is expected to be one of two things:
  *
  *    - If a data response is expected on success, a struct type decodable via DataModel::Decode which has GetClusterId() and
  * GetCommandId() methods.  A ClusterName::Commands::ResponseCommandName::DecodableType is typically used.
- *    - If a status response is expected on success, a DataModel::NullObjectType.
+ *    - If a status response is expected on success, DataModel::NullObjectType.
  *
  */
-template <typename ResponseObjectT = app::DataModel::NullObjectType, typename RequestObjectT>
-CHIP_ERROR InvokeCommandRequest(Messaging::ExchangeManager * aExchangeMgr, SessionHandle sessionHandle, chip::EndpointId endpointId,
-                                const RequestObjectT & requestCommandData,
-                                typename TypedCommandCallback<ResponseObjectT>::OnSuccessCallbackType onSuccessCb,
-                                typename TypedCommandCallback<ResponseObjectT>::OnErrorCallbackType onErrorCb)
+template <typename RequestObjectT>
```

Comment:
Would it be worthwhile adding ResponseObjectT, but initialized to a default of `RequestObjectT::ResponseType`?

That way, you can easily over-ride that if you need to, without needing to create custom inherited types.

Reply:
We could do that, but then the overriding invocations would need to provide both the request and response type explicitly.  Is that less typing/boilerplate than the inherited type?  Also note that this is a use case that fundamentally I'm not sure we actually want to support in a "it's clear how to do this" way, and that is not needed so far anywhere except these two tests that basically send a "spec-violating" (in the sense that it's not allowed by the XML) response.

Thoughts?

Reply:
My suggestion was mostly to provide an escape hatch in the cases where the ZAP inferred response type was incorrect. Thinking about this a bit more, the correct solution there would be just force people to fix the XML than to permit overrides here, so perhaps, this shouldn't be made available at the risk of abuse.

Reply:
Yes, if the ZAP info is wrong adding the right info to the XML is the thing to do, I agree.